lib/spack/spack/cmd/config.py

```diff
@@ -135,11 +164,150 @@ def config_list(args):
+        override = False
+        for idx, name in enumerate(components[:-1]):
+            # First handle double colons in constructing path
+            colon = '::' if override else ':' if path else ''
+            path += colon + name
+            if getattr(name, 'override', False):
+                override = True
+            else:
+                override = False
+
+            # Test whether there is an existing value at this level
+            existing = spack.config.get(path, scope=scope)
+
+            if existing is None:
+                has_existing_value = False
+                # We've nested further than existing config, so we need the
+                # type information for validation to know how to handle bare
+                # values appended to lists.
+                existing = spack.config.type_of(path)
+
+                # construct value from this point down
+                value = syaml.load_config(components[-1])
+                for component in reversed(components[idx + 1:-1]):
+                    value = {component: value}
+                break
+
+        if has_existing_value:
+            path, _, value = args.value.rpartition(':')
+            value = syaml.load_config(value)
+            existing = spack.config.get(path, scope=scope)
+
+        # append values to lists appropriately
+        if isinstance(existing, list) and not isinstance(value, list):
+            value = [value]
+
+        # merge value into existing
+        new = spack.config.merge_yaml(existing, value)
+
+        if re.match(r'env.*', scope or ''):
+            e = ev.get_env(args, 'config add')
+            e.set_config(path, new)
+        else:
+            spack.config.set(path, new, scope=scope)
+
+
+def config_remove(args):
+    """Remove the given configuration from the specified config scope
+
+    This is a stateful operation that edits the config files under the hood"""
+    scope, _ = _get_scope_and_section(args)
+
+    e = ev.get_env(args, 'config remove')
+    if e:
+        import ruamel.yaml as yaml
+        tty.msg(str(getattr(e.yaml, yaml.comments.Comment.attrib, None)))
+
+    path, _, value = args.value.rpartition(':')
+    existing = spack.config.get(path, scope=scope)
+
+    if not isinstance(existing, (list, dict)):
+        path, _, value = path.rpartition(':')
+        existing = spack.config.get(path, scope=scope)
+
+    value = syaml.load(value)
+
+    if isinstance(existing, list):
+        values = value if isinstance(value, list) else [value]
+        for v in values:
+            existing.remove(v)
+    elif isinstance(existing, dict):
+        existing.pop(value, None)
+    else:
+        # This should be impossible to reach
+        raise spack.config.ConfigError('Config has nested non-dict values')
+
+    if re.match(r'env.*', scope or ''):
+        e = ev.get_env(args, 'config remove')
+        e.set_config(path, existing)
+    else:
+        spack.config.set(path, existing, scope=scope)
```

Comment:
I'm a little worried about this code as well, as it seems *very* tied to the internals of the config system.  Can this stuff (particularly the path handling and YAML code) be moved into `spack.config`?